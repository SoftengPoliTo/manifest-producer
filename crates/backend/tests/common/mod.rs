use insta::{assert_snapshot, with_settings};
use manifest_producer_backend::{
    analyse::analyse_functions,
    detect::function_detection,
    entry::find_root_nodes,
    inspect::{inspect_binary, parse_elf, read_elf},
};
use serde_json::Value;
use std::{collections::BTreeMap, env::temp_dir, fs, path::Path};

pub fn run_analysis_test(binary_path: &str, test_name: &str) {
    let output_path = setup_test_environment(test_name);

    let elf_buffer = read_elf(binary_path).unwrap();
    let elf = parse_elf(&elf_buffer).unwrap();

    let info = inspect_binary(&elf, binary_path, &output_path).unwrap();

    let mut detected_functions = function_detection(&elf, &info.language, &output_path).unwrap();
    analyse_functions(
        &elf,
        &elf_buffer,
        &mut detected_functions,
        &info.language,
        &output_path,
    )
    .unwrap();

    let inspect_json = read_and_sort_json(Path::new(&output_path).join("json/basic_info.json"));
    let analyse_json = read_and_sort_json(Path::new(&output_path).join("json/functions_list.json"));

    let root_nodes = find_root_nodes(binary_path, &info.language, &detected_functions).unwrap();
    let mut sorted_root_nodes = root_nodes.clone();
    sorted_root_nodes.sort();
    let root_nodes_json = serde_json::to_string_pretty(&sorted_root_nodes).unwrap();

    with_settings!(
        {
            snapshot_path => format!("../snapshots/{}/", test_name),
            prepend_module_to_snapshot => false,
        },
        {
            assert_snapshot!("basic_info", inspect_json);
            assert_snapshot!("functions_list", analyse_json);
            assert_snapshot!("root_nodes", root_nodes_json);
        }
    );
}

fn setup_test_environment(test_name: &str) -> String {
    let test_dir = temp_dir().join(test_name);
    let json_dir = test_dir.join("json");

    for dir in [&test_dir, &json_dir] {
        if !dir.exists() {
            fs::create_dir_all(dir).unwrap();
        }
    }

    test_dir.to_str().unwrap().to_owned()
}

fn read_and_sort_json<P: AsRef<Path>>(path: P) -> String {
    let json_content = fs::read_to_string(&path).unwrap();
    let parsed_json: Value = serde_json::from_str(&json_content).unwrap();

    match parsed_json {
        Value::Object(map) => {
            let sorted_map: BTreeMap<_, _> = map.into_iter().collect();
            serde_json::to_string_pretty(&sorted_map).unwrap()
        }
        _ => json_content,
    }
}
