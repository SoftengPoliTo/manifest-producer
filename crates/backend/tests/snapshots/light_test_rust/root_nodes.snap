---
source: crates/backend/tests/common/mod.rs
expression: root_nodes_json
snapshot_kind: text
---
{
  "name": "main",
  "start_addr": 4732128,
  "end_addr": 4733575,
  "invocation_entry": 0,
  "jmp": 0,
  "children": [
    "std::sys::pal::unix::stack_overflow::imp::make_handler::h8dd41e2509d66e05",
    "std::thread::Thread::new_inner::h9da9ce7e0b97ec5f",
    "std::thread::set_current::h1e3fe36e53a71c47",
    "std::sys::backtrace::__rust_begin_short_backtrace::hbcf0ccfa82933b96",
    "std::io::stdio::attempt_print_to_stderr::h10a68257d4fae583",
    "std::sys::sync::once::futex::Once::call::h569431cc4b2ed5fa",
    "std::sys::exit_guard::unique_thread_exit::h28d77dc0bd6dcd93",
    "std::io::Write::write_fmt::h8c2535c68af369a5",
    "core::ptr::drop_in_place<core::result::Result<(),std::io::error::Error>>::hc6129a43312b9fc1",
    "std::sys::pal::unix::abort_internal::h1c3bfb6c2fa864f5",
    "__rust_panic_cleanup",
    "core::ops::function::FnOnce::call_once::h480eb1908e77135b"
  ],
  "disassembly": "0x4834e0:\tpush\t%r15\n0x4834e2:\tpush\t%r14\n0x4834e4:\tpush\t%r13\n0x4834e6:\tpush\t%r12\n0x4834e8:\tpush\t%rbx\n0x4834e9:\tsub\t$0x130, %rsp\n0x4834f0:\tmov\t%rsi, %rbx\n0x4834f3:\tmovzx\t0x17b992(%rip), %eax\n0x4834fa:\tmovsxd\t%edi, %r15\n0x4834fd:\tmov\t$0, 0x10(%rsp)\n0x483506:\tmovaps\t0x157c83(%rip), %xmm0\n0x48350d:\tmovups\t%xmm0, 0x18(%rsp)\n0x483512:\tlea\t0x10(%rsp), %r14\n0x483517:\tlea\t0x150194(%rip), %r12\n0x48351e:\tlea\t0x14b11b(%rip), %r13\n0x483525:\tnop\t%cs:(%rax, %rax)\n0x483530:\tmov\t$3, %esi\n0x483535:\tmov\t%r14, %rdi\n0x483538:\txor\t%edx, %edx\n0x48353a:\tcall\t*%r12\t(Register Offset-GOT)\n0x48353d:\tcmp\t$-1, %eax\n0x483540:\tjne\t0x48360a\n0x483546:\tcall\t*%r13\t(Register Offset-GOT)\n0x483549:\tmov\t(%rax), %eax\n0x48354b:\tcmp\t$4, %eax\n0x48354e:\tje\t0x483530\n0x483550:\tcmp\t$0x16, %eax\n0x483553:\tja\t0x483916\n0x483559:\tmov\t$0x401800, %ecx\n0x48355e:\tbt\t%eax, %ecx\n0x483561:\tjae\t0x483916\n0x483567:\txor\t%edi, %edi\n0x483569:\tmov\t$1, %esi\n0x48356e:\txor\t%eax, %eax\n0x483570:\tcall\t*0x1cba42(%rip)\t(Register Offset-GOT)\n0x483576:\tcmp\t$-1, %eax\n0x483579:\tjne\t0x4835a5\n0x48357b:\tcall\t*0x1cb9f7(%rip)\t(Register Offset-GOT)\n0x483581:\tcmp\t$9, (%rax)\n0x483584:\tjne\t0x4835a5\n0x483586:\tlea\t0x171a53(%rip), %rdi\n0x48358d:\tmov\t$2, %esi\n0x483592:\txor\t%edx, %edx\n0x483594:\txor\t%eax, %eax\n0x483596:\tcall\t*0x1cb9fc(%rip)\t(Register Offset-GOT)\n0x48359c:\tcmp\t$-1, %eax\n0x48359f:\tje\t0x483916\n0x4835a5:\tmov\t$1, %edi\n0x4835aa:\tmov\t$1, %esi\n0x4835af:\txor\t%eax, %eax\n0x4835b1:\tcall\t*0x1cba01(%rip)\t(Register Offset-GOT)\n0x4835b7:\tcmp\t$-1, %eax\n0x4835ba:\tjne\t0x4835e6\n0x4835bc:\tcall\t*0x1cb9b6(%rip)\t(Register Offset-GOT)\n0x4835c2:\tcmp\t$9, (%rax)\n0x4835c5:\tjne\t0x4835e6\n0x4835c7:\tlea\t0x171a12(%rip), %rdi\n0x4835ce:\tmov\t$2, %esi\n0x4835d3:\txor\t%edx, %edx\n0x4835d5:\txor\t%eax, %eax\n0x4835d7:\tcall\t*0x1cb9bb(%rip)\t(Register Offset-GOT)\n0x4835dd:\tcmp\t$-1, %eax\n0x4835e0:\tje\t0x483916\n0x4835e6:\tmov\t$2, %edi\n0x4835eb:\tmov\t$1, %esi\n0x4835f0:\txor\t%eax, %eax\n0x4835f2:\tcall\t*0x1cb9c0(%rip)\t(Register Offset-GOT)\n0x4835f8:\tcmp\t$-1, %eax\n0x4835fb:\tjne\t0x48367c\n0x4835fd:\tcall\t*0x1cb975(%rip)\t(Register Offset-GOT)\n0x483603:\tcmp\t$9, (%rax)\n0x483606:\tje\t0x48365d\n0x483608:\tjmp\t0x48367c\n0x48360a:\ttest\t$0x20, 0x16(%rsp)\n0x48360f:\tje\t0x483630\n0x483611:\tlea\t0x1719c8(%rip), %rdi\n0x483618:\tmov\t$2, %esi\n0x48361d:\txor\t%edx, %edx\n0x48361f:\txor\t%eax, %eax\n0x483621:\tcall\t*0x1cb971(%rip)\t(Register Offset-GOT)\n0x483627:\tcmp\t$-1, %eax\n0x48362a:\tje\t0x483916\n0x483630:\ttest\t$0x20, 0x1e(%rsp)\n0x483635:\tje\t0x483656\n0x483637:\tlea\t0x1719a2(%rip), %rdi\n0x48363e:\tmov\t$2, %esi\n0x483643:\txor\t%edx, %edx\n0x483645:\txor\t%eax, %eax\n0x483647:\tcall\t*0x1cb94b(%rip)\t(Register Offset-GOT)\n0x48364d:\tcmp\t$-1, %eax\n0x483650:\tje\t0x483916\n0x483656:\ttest\t$0x20, 0x26(%rsp)\n0x48365b:\tje\t0x48367c\n0x48365d:\tlea\t0x17197c(%rip), %rdi\n0x483664:\tmov\t$2, %esi\n0x483669:\txor\t%edx, %edx\n0x48366b:\txor\t%eax, %eax\n0x48366d:\tcall\t*0x1cb925(%rip)\t(Register Offset-GOT)\n0x483673:\tcmp\t$-1, %eax\n0x483676:\tje\t0x483916\n0x48367c:\tmov\t$1, %esi\n0x483681:\tmov\t$0xd, %edi\n0x483686:\tcall\t*0x1cb7fc(%rip)\t(Register Offset-GOT)\n0x48368c:\tcmp\t$-1, %rax\n0x483690:\tje\t0x48391c\n0x483696:\tmov\t$0x1e, %edi\n0x48369b:\tcall\t*0x1cb667(%rip)\t(Register Offset-GOT)\n0x4836a1:\tmov\t%rax, 0x1cc3f8(%rip)\n0x4836a8:\tmov\t0x1cc3f1(%rip), %rax\n0x4836af:\txorps\t%xmm0, %xmm0\n0x4836b2:\tmovups\t%xmm0, %fs:0xffffffffffffff10\n0x4836bb:\tmovaps\t%xmm0, 0x90(%rsp)\n0x4836c3:\tmovaps\t%xmm0, 0x80(%rsp)\n0x4836cb:\tmovaps\t%xmm0, 0x70(%rsp)\n0x4836d0:\tmovaps\t%xmm0, 0x60(%rsp)\n0x4836d5:\tmovaps\t%xmm0, 0x50(%rsp)\n0x4836da:\tmovaps\t%xmm0, 0x40(%rsp)\n0x4836df:\tmovaps\t%xmm0, 0x30(%rsp)\n0x4836e4:\tmovaps\t%xmm0, 0x20(%rsp)\n0x4836e9:\tmovaps\t%xmm0, 0x10(%rsp)\n0x4836ee:\tmov\t$0, 0xa0(%rsp)\n0x4836fa:\tlea\t0x10(%rsp), %rdx\n0x4836ff:\tmov\t$0xb, %edi\n0x483704:\txor\t%esi, %esi\n0x483706:\tcall\t*0x1cb7dc(%rip)\t(Register Offset-GOT)\n0x48370c:\tcmp\t$0, 0x10(%rsp)\n0x483712:\tjne\t0x483760\n0x483714:\tmovzx\t0x1cc395(%rip), %eax\n0x48371b:\ttest\t%al, %al\n0x48371d:\tjne\t0x483737\n0x48371f:\tmov\t$1, 0x1cc38a(%rip)\n0x483726:\tmov\t$1, %edi\n0x48372b:\tcall\t0x56fb00   <std::sys::pal::unix::stack_overflow::imp::make_handler::h8dd41e2509d66e05>\n0x483730:\tmov\t%rax, 0x1cc371(%rip)\n0x483737:\tmov\t$0x8000004, 0x98(%rsp)\n0x483742:\tlea\t0xec6f7(%rip), %rax\n0x483749:\tmov\t%rax, 0x10(%rsp)\n0x48374e:\tlea\t0x10(%rsp), %rsi\n0x483753:\tmov\t$0xb, %edi\n0x483758:\txor\t%edx, %edx\n0x48375a:\tcall\t*0x1cb788(%rip)\t(Register Offset-GOT)\n0x483760:\tlea\t0x10(%rsp), %rdx\n0x483765:\tmov\t$7, %edi\n0x48376a:\txor\t%esi, %esi\n0x48376c:\tcall\t*0x1cb776(%rip)\t(Register Offset-GOT)\n0x483772:\tcmp\t$0, 0x10(%rsp)\n0x483778:\tjne\t0x4837c6\n0x48377a:\tmovzx\t0x1cc32f(%rip), %eax\n0x483781:\ttest\t%al, %al\n0x483783:\tjne\t0x48379d\n0x483785:\tmov\t$1, 0x1cc324(%rip)\n0x48378c:\tmov\t$1, %edi\n0x483791:\tcall\t0x56fb00   <std::sys::pal::unix::stack_overflow::imp::make_handler::h8dd41e2509d66e05>\n0x483796:\tmov\t%rax, 0x1cc30b(%rip)\n0x48379d:\tmov\t$0x8000004, 0x98(%rsp)\n0x4837a8:\tlea\t0xec691(%rip), %rax\n0x4837af:\tmov\t%rax, 0x10(%rsp)\n0x4837b4:\tlea\t0x10(%rsp), %rsi\n0x4837b9:\tmov\t$7, %edi\n0x4837be:\txor\t%edx, %edx\n0x4837c0:\tcall\t*0x1cb722(%rip)\t(Register Offset-GOT)\n0x4837c6:\tmov\t%r15, 0x1cc2b3(%rip)\n0x4837cd:\tmov\t%rbx, 0x1cc2b4(%rip)\n0x4837d4:\tmov\t$0, 0x10(%rsp)\n0x4837dd:\tlea\t0x10(%rsp), %rdi\n0x4837e2:\tcall\t0x56ffe0   <std::thread::Thread::new_inner::h9da9ce7e0b97ec5f>\n0x4837e7:\tmov\t%rax, %rdi\n0x4837ea:\tcall\t0x5700c0   <std::thread::set_current::h1e3fe36e53a71c47>\n0x4837ef:\tlea\t-0xa716(%rip), %rsi\n0x4837f6:\tlea\t0x110(%rsp), %rdi\n0x4837fe:\tcall\t0x436c70   <std::sys::backtrace::__rust_begin_short_backtrace::hbcf0ccfa82933b96>\n0x483803:\tmovabs\t$9223372036854775809, %rax\n0x48380d:\txor\t%ebx, %ebx\n0x48380f:\tcmp\t%rax, 0x110(%rsp)\n0x483817:\tje\t0x4838bd\n0x48381d:\tmovups\t0x110(%rsp), %xmm0\n0x483825:\tmovups\t0x120(%rsp), %xmm1\n0x48382d:\tmovaps\t%xmm1, 0xc0(%rsp)\n0x483835:\tmovaps\t%xmm0, 0xb0(%rsp)\n0x48383d:\tlea\t0xb0(%rsp), %rax\n0x483845:\tmov\t%rax, 0x100(%rsp)\n0x48384d:\tlea\t-0x26894(%rip), %rax\n0x483854:\tmov\t%rax, 0x108(%rsp)\n0x48385c:\tlea\t0x1bd91d(%rip), %rax\n0x483863:\tmov\t%rax, 0x10(%rsp)\n0x483868:\tmov\t$2, 0x18(%rsp)\n0x483871:\tmov\t$0, 0x30(%rsp)\n0x48387a:\tlea\t0x100(%rsp), %rax\n0x483882:\tmov\t%rax, 0x20(%rsp)\n0x483887:\tmov\t$1, 0x28(%rsp)\n0x483890:\tlea\t0x10(%rsp), %rdi\n0x483895:\tcall\t0x5984b0   <std::io::stdio::attempt_print_to_stderr::h10a68257d4fae583>\n0x48389a:\tmov\t0xb0(%rsp), %rax\n0x4838a2:\tshl\t$1, %rax\n0x4838a5:\ttest\t%rax, %rax\n0x4838a8:\tje\t0x4838b8\n0x4838aa:\tmov\t0xb8(%rsp), %rdi\n0x4838b2:\tcall\t*0x1cb738(%rip)\t(Register Offset-GOT)\n0x4838b8:\tmov\t$1, %ebx\n0x4838bd:\tmov\t0x1cc139(%rip), %eax\n0x4838c3:\tcmp\t$3, %eax\n0x4838c6:\tje\t0x4838fe\n0x4838c8:\tmov\t$1, 0xf(%rsp)\n0x4838cd:\tlea\t0xf(%rsp), %rax\n0x4838d2:\tmov\t%rax, 0xd0(%rsp)\n0x4838da:\tlea\t0x1cc11b(%rip), %rdi\n0x4838e1:\tlea\t0x1c7448(%rip), %rcx\n0x4838e8:\tlea\t0x1c6fe9(%rip), %r8\n0x4838ef:\tlea\t0xd0(%rsp), %rdx\n0x4838f7:\txor\t%esi, %esi\n0x4838f9:\tcall\t0x417580   <std::sys::sync::once::futex::Once::call::h569431cc4b2ed5fa>\n0x4838fe:\tcall\t0x5701f0   <std::sys::exit_guard::unique_thread_exit::h28d77dc0bd6dcd93>\n0x483903:\tmov\t%ebx, %eax\n0x483905:\tadd\t$0x130, %rsp\n0x48390c:\tpop\t%rbx\n0x48390d:\tpop\t%r12\n0x48390f:\tpop\t%r13\n0x483911:\tpop\t%r14\n0x483913:\tpop\t%r15\n0x483915:\tret\t\n0x483916:\tcall\t*0x1cb4fc(%rip)\t(Register Offset-GOT)\n0x48391c:\tlea\t0x1c7b7d(%rip), %rax\n0x483923:\tmov\t%rax, 0x10(%rsp)\n0x483928:\tmov\t$1, 0x18(%rsp)\n0x483931:\tlea\t0xd0(%rsp), %rdi\n0x483939:\tmov\t%rdi, 0x20(%rsp)\n0x48393e:\txorps\t%xmm0, %xmm0\n0x483941:\tmovups\t%xmm0, 0x28(%rsp)\n0x483946:\tlea\t0x10(%rsp), %rsi\n0x48394b:\tcall\t0x56f960   <std::io::Write::write_fmt::h8c2535c68af369a5>\n0x483950:\tmov\t%rax, %rdi\n0x483953:\tcall\t0x56fa70   <core::ptr::drop_in_place<core::result::Result<(),std::io::error::Error>>::hc6129a43312b9fc1>\n0x483958:\tcall\t0x56faf0   <std::sys::pal::unix::abort_internal::h1c3bfb6c2fa864f5>\n0x48395d:\tmov\t%rax, %rbx\n0x483960:\tmov\t0xb0(%rsp), %rax\n0x483968:\tshl\t$1, %rax\n0x48396b:\ttest\t%rax, %rax\n0x48396e:\tje\t0x483988\n0x483970:\tmov\t0xb8(%rsp), %rdi\n0x483978:\tcall\t*0x1cb672(%rip)\t(Register Offset-GOT)\n0x48397e:\tjmp\t0x483988\n0x483980:\tjmp\t0x483a60\n0x483985:\tmov\t%rax, %rbx\n0x483988:\tmov\t%rbx, %rdi\n0x48398b:\tcall\t0x561990   <__rust_panic_cleanup>\n0x483990:\tdec\t0x1cc1b0(%rip)\n0x483998:\tmov\t%rax, %r14\n0x48399b:\tdec\t%fs:0xffffffffffffff00\n0x4839a4:\tmov\t%rdx, %r15\n0x4839a7:\tmov\t$0, %fs:0xffffffffffffff08\n0x4839b0:\tmov\t(%rdx), %rax\n0x4839b3:\ttest\t%rax, %rax\n0x4839b6:\tje\t0x4839bd\n0x4839b8:\tmov\t%r14, %rdi\n0x4839bb:\tcall\t*%rax\t(Register Offset-GOT)\n0x4839bd:\tmov\t$0x65, %ebx\n0x4839c2:\tcmp\t$0, 8(%r15)\n0x4839c7:\tje\t0x4838bd\n0x4839cd:\tmov\t%r14, %rdi\n0x4839d0:\tcall\t*0x1cb61a(%rip)\t(Register Offset-GOT)\n0x4839d6:\tjmp\t0x4838bd\n0x4839db:\tmov\t%rax, %rbx\n0x4839de:\tcmp\t$0, 8(%r15)\n0x4839e3:\tje\t0x4839ee\n0x4839e5:\tmov\t%r14, %rdi\n0x4839e8:\tcall\t*0x1cb602(%rip)\t(Register Offset-GOT)\n0x4839ee:\tmov\t%rbx, %rdi\n0x4839f1:\tcall\t0x561990   <__rust_panic_cleanup>\n0x4839f6:\tdec\t0x1cc14a(%rip)\n0x4839fe:\tdec\t%fs:0xffffffffffffff00\n0x483a07:\tmov\t$0, %fs:0xffffffffffffff08\n0x483a10:\tlea\t0x1c6ed9(%rip), %rax\n0x483a17:\tmov\t%rax, 0xd0(%rsp)\n0x483a1f:\tmov\t$1, 0xd8(%rsp)\n0x483a2b:\tlea\t0xd0(%rsp), %rdi\n0x483a33:\tmov\t%rdi, 0xe0(%rsp)\n0x483a3b:\txorps\t%xmm0, %xmm0\n0x483a3e:\tmovups\t%xmm0, 0xe8(%rsp)\n0x483a46:\tlea\t0xd0(%rsp), %rsi\n0x483a4e:\tcall\t0x56f960   <std::io::Write::write_fmt::h8c2535c68af369a5>\n0x483a53:\tmov\t%rax, %rdi\n0x483a56:\tcall\t0x56fa70   <core::ptr::drop_in_place<core::result::Result<(),std::io::error::Error>>::hc6129a43312b9fc1>\n0x483a5b:\tcall\t0x56faf0   <std::sys::pal::unix::abort_internal::h1c3bfb6c2fa864f5>\n0x483a60:\tmov\t%rax, %rdi\n0x483a63:\tcall\t0x561990   <__rust_panic_cleanup>\n0x483a68:\tdec\t0x1cc0d8(%rip)\n0x483a70:\tdec\t%fs:0xffffffffffffff00\n0x483a79:\tmov\t$0, %fs:0xffffffffffffff08\n0x483a82:\tcall\t0x5701e0   <core::ops::function::FnOnce::call_once::h480eb1908e77135b>\n",
  "syscall": false,
  "syscall_info": null
}
